File: C++/20. Valid Parentheses.cpp
#include <bits/stdc++.h>

using namespace std;

bool isOpen(char ch)
{
    return (ch == '[' or ch == '{' or ch == '(');
}

bool isMatch(char open, char close)
{
    switch (open)
    {
    case '{':
        return (close == '}');
    case '[':
        return (close == ']');
    case '(':
        return (close == ')');
    }
    return false;
}

bool isValid(string s)
{
    stack<char> st;

    for (char c : s)
    {
        if (isOpen(c))
            st.push(c);
        else if (!st.empty() and isMatch(st.top(), c))
            st.pop();
        else
            return false;
    }

    return st.empty();
}

File: Java/20. Valid Parentheses.java

import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (isOpen(c))
                stack.add(c);
            else if (stack.isEmpty() || !isMatchingPair(stack.pop(), c))
                return false;
        }

        return stack.isEmpty();
    }

    private boolean isMatchingPair(char open, char close) {
        return switch (open) {
            case '(' -> close == ')';
            case '{' -> close == '}';
            case '[' -> close == ']';
            default -> false;
        };
    }

    private boolean isOpen(char c) {
        return (c == '(' || c == '{' || c == '[');
    }
}